types/react/index.d.ts

```diff
@@ -2633,30 +2658,76 @@ declare namespace React {
-type MergePropTypes<P, T> = P & Pick<T, Exclude<keyof T, keyof P>>;
-
-// Any prop that has a default prop becomes optional, but its type is unchanged
-// Undeclared default props are augmented into the resulting allowable attributes
-// If declared props have indexed properties, ignore default props entirely as keyof gets widened
-// Wrap in an outer-level conditional type to allow distribution over props that are unions
-type Defaultize<P, D> = P extends any
-    ? string extends keyof P ? P :
-        & Pick<P, Exclude<keyof P, keyof D>>
-        & Partial<Pick<P, Extract<keyof P, keyof D>>>
-        & Partial<Pick<D, Exclude<keyof D, keyof P>>>
-    : never;
-
-type ReactManagedAttributes<C, P> = C extends { propTypes: infer T; defaultProps: infer D; }
-    ? Defaultize<MergePropTypes<P, PropTypes.InferProps<T>>, D>
-    : C extends { propTypes: infer T; }
-        ? MergePropTypes<P, PropTypes.InferProps<T>>
-        : C extends { defaultProps: infer D; }
-            ? Defaultize<P, D>
-            : P;
+    const Invalid_propTypes: unique symbol;
+    /**
+     * Dummy interface that is intersected into the props type of a component on
+     * use to generate an error if the component's `propTypes` static property
+     * is invalid.  In most cases, you can get more information about the
+     * problem with the `propTypes` by trying to assign the component to a
+     * variable of type `React.ComponentType<P>` where `P` is the appropriate
+     * props type.  You can bypass the type checking of `propTypes` by defining
+     * a static property named `bypassPropTypesTypecheck`.
+     */
+    interface Invalid_propTypes<P, T> {
+        [Invalid_propTypes]: undefined;
+    }
+
+    const Invalid_defaultProps: unique symbol;
+    /**
+     * Dummy interface that is intersected into the props type of a component on
+     * use to generate an error if the component's `defaultProps` static
+     * property is invalid.  In most cases, you can get more information about
+     * the problem with the `defaultProps` by trying to assign the component to
+     * a variable of type `React.ComponentType<P>` where `P` is the appropriate
+     * props type.  You can bypass the type checking of `defaultProps` by
+     * defining a static property named `bypassDefaultPropsTypecheck`.
+     */
+    interface Invalid_defaultProps<P, D> {
+        [Invalid_defaultProps]: undefined;
+    }
+
+    // For a component class, the P passed to JSX.LibraryManagedAttributes is
+    // based on the `props` property and includes `children` regardless of
+    // whether the original props type (the `P` type argument to
+    // React.Component) does.  However, it's common for component classes not
+    // to list `children` in the `propTypes` if the original props type did not
+    // include `children`.  We accommodate that here by making the validator for
+    // `children` optional.
```

Comment:
The React types has an open problem (that `{}` in `ReactFragment` which makes `ReactNode` useless) that will have to change that, one way or another. I have an open PR that does it, but maybe there are better ideas on how to do it.

Reply:
Are you suggesting a change to this pull request?